README.md
# defy

[![Travis branch](https://img.shields.io/travis/MrSlide/defy/master.svg)](https://travis-ci.org/MrSlide/defy)
[![Coverage Status](https://img.shields.io/coveralls/MrSlide/defy/master.svg)](https://coveralls.io/github/MrSlide/defy?branch=master)
[![GitHub issues](https://img.shields.io/github/issues/mrslide/defy.svg)](https://github.com/MrSlide/defy)
[![Dependencies](https://img.shields.io/david/mrslide/defy.svg)](https://www.npmjs.com/package/defy)
[![npm](https://img.shields.io/npm/v/defy.svg)](https://www.npmjs.com/package/defy)
[![npm](https://img.shields.io/npm/dt/defy.svg)](https://www.npmjs.com/package/defy)
[![Standard](https://img.shields.io/badge/code_style-standard-brightgreen.svg)](https://standardjs.com)
[![license](https://img.shields.io/github/license/mrslide/defy.svg)](https://github.com/MrSlide/defy)

Small JavaScript library for data validation with flexibility in mind. It won't decide how you do form validation or present error messages. You simply provide some value to validate, and it will tell you if it is valid or not.



## Install

```
npm install defy
```



## Usage

Defy is a UMD and ES6 module.

```
// Loading the module in ES6 style
import * as defy from 'defy'

// It is also possible to only load the required validations when loading the module in ES6 style (this is the preferred way when possible)
import {required, password} from 'defy'
import {date as defyDate} from 'defy'

// Some module bundlers will give preference to the UMD module instead of the ES6 module.
// Force loading the ES6 module instead by including the index.js file in the import path.
import {date as defyDate} from 'defy/index.js'

// Loading the module in CommonJS style
const defy = require('defy')
```



## Static Public Methods

### age(dob: String[, minAge: Number = 0[, maxAge: Number]]): Boolean | Number

Verify if someone's age falls within a minimum and maximum age values and get the age value.

#### Params

| Name                     | Type                     | Description                                                                                                     |
|--------------------------|--------------------------|-----------------------------------------------------------------------------------------------------------------|
| dob                      | String                   | A valid RFC3339 date string representing the date of birth.                                                     |
| minAge                   | Number                   | The minimum accepted age value.                                                                                 |
| maxAge                   | Number                   | The maximum accepted age value.                                                                                 |

#### Return

Boolean | Number - Returns `false` if the minimum or maximum age are not valid, or the calculated age if valid.

#### Example

```
var isAgeValid = defy.age('1999-02-22', 18)
var userAge = defy.age('1979-06-12')
```



## Static Public Methods

### alphaNumeric(value: String): Boolean

Verify if a given value only contains alpha-numeric characters.

#### Params

| Name                     | Type                     | Description                                                                                                     |
|--------------------------|--------------------------|-----------------------------------------------------------------------------------------------------------------|
| value                    | String                   | The value to be validated.                                                                                      |

#### Return

Boolean

#### Example

```
var isValueAlphaNumeric = defy.alphaNumeric('Abc123')
```



### cleanValue(value: Mixed): Boolean

Clean up a given value to remove extra spaces that may cause false validation results.

#### Params

| Name                     | Type                     | Description                                                                                                     |
|--------------------------|--------------------------|-----------------------------------------------------------------------------------------------------------------|
| value                    | Mixed                    | The value to be validated.                                                                                      |

#### Return

Boolean

#### Throw

##### TypeError

If trying to clean a value that cannot be converted to a string.

#### Example

```
var cleanedValue = defy.cleanValue('  Example   ')
```



### color(value: String): Boolean

Verify if a given value is a valid hexadecimal format colour with 3 or 6 characters and preceded by the hash sign.

#### Params

| Name                     | Type                     | Description                                                                                                     |
|--------------------------|--------------------------|-----------------------------------------------------------------------------------------------------------------|
| value                    | String                   | The value to be validated.                                                                                      |

#### Return

Boolean

#### Example

```
var isColourValid = defy.color('#FFFFFF')
```



### creditcard(value: String): Boolean

Verify if a given value is a valid credit card number.

#### Params

| Name                     | Type                     | Description                                                                                                     |
|--------------------------|--------------------------|-----------------------------------------------------------------------------------------------------------------|
| value                    | String                   | The value to be validated.                                                                                      |

#### Return

Boolean

#### Example

```
var isCCValid = defy.creditcard('378282246310005')
```



### date(value: String): Boolean

Verify if a given value is a valid date.

#### Params

| Name                     | Type                     | Description                                                                                                     |
|--------------------------|--------------------------|-----------------------------------------------------------------------------------------------------------------|
| value                    | String                   | The value to be validated.                                                                                      |

#### Return

Boolean

#### Example

```
var isDateValid = defy.date('2017-05-31')
```



### dateExists(value: String): Boolean

Verify if a given date exists. The value must be a valid date string and pass the `defy.date()` validation.

#### Params

| Name                     | Type                     | Description                                                                                                     |
|--------------------------|--------------------------|-----------------------------------------------------------------------------------------------------------------|
| value                    | String                   | The value to be validated.                                                                                      |

#### Return

Boolean

#### Example

```
var dateExists = defy.dateExists('2017-05-31')
```



### email(value: String): Boolean

Verify if a given value is a valid email address.

#### Params

| Name                     | Type                     | Description                                                                                                     |
|--------------------------|--------------------------|-----------------------------------------------------------------------------------------------------------------|
| value                    | String                   | The value to be validated.                                                                                      |

#### Return

Boolean

#### Example

```
var isEmailValid = defy.email('user@world.net');
```



### match(valueA: Mixed, valueB: Mixed): Boolean

Verify if two given values match. Both values will be treated as strings.

#### Params

| Name                     | Type                     | Description                                                                                                     |
|--------------------------|--------------------------|-----------------------------------------------------------------------------------------------------------------|
| valueA                   | Mixed                    | The first value to match.                                                                                       |
| valueB                   | Mixed                    | The second value, which will be compared with the first.                                                        |

#### Return

Boolean

#### Example

```
var doValuesMatch = defy.match(1, '1')
```



### max(value: Number | String, maxValue: Number | String): Boolean

Verify if a given value does not exceed another given value. Both values will be treated as strings.

#### Params

| Name                     | Type                     | Description                                                                                                     |
|--------------------------|--------------------------|-----------------------------------------------------------------------------------------------------------------|
| value                    | Number | String          | The value to be validated.                                                                                      |
| maxValue                 | Mixed                    | The maximum allowed for the given value.                                                                        |

#### Return

Boolean

#### Example

```
var isValueLessThanEleven = defy.max(5, 10)
```



### maxlength(value: Number | String, maxLen: Number): Boolean

Verify if a given value's length does not exceed the given maximum length. For strings this will validate the number of characters. For numbers, this will validate the number of digits, including the decimal places, and the decimal mark.

#### Params

| Name                     | Type                     | Description                                                                                                     |
|--------------------------|--------------------------|-----------------------------------------------------------------------------------------------------------------|
| value                    | Number | String          | The value to be validated.                                                                                      |
| maxLen                   | Number                   | The maximum allowed for the given value.                                                                        |

#### Return

Boolean

#### Example

```
var isValueNotTooLong = defy.maxlength('Example', 10)
```



### min(value: Number | String, minValue: Number | String): Boolean

Verify if a given value is not inferior to another given value. Both values will be treated as strings.

#### Params

| Name                     | Type                     | Description                                                                                                     |
|--------------------------|--------------------------|-----------------------------------------------------------------------------------------------------------------|
| value                    | Number | String          | The value to be validated.                                                                                      |
| minValue                 | Number | String          | The minimum allowed for the given value.                                                                        |

#### Return

Boolean

#### Example

```
var isValueGreaterThanZero = defy.min(5, 0)
```



### minlength(value: Number | String, minLen: Number): Boolean

Verify if a given value's length is at least the given minimum length. For strings this will validate the number of characters. For numbers, this will validate the number of digits, including the decimal places, and the decimal mark.

#### Params

| Name                     | Type                     | Description                                                                                                     |
|--------------------------|--------------------------|-----------------------------------------------------------------------------------------------------------------|
| value                    | Number | String          | The value to be validated.                                                                                      |
| minLen                   | Number                   | The minimum length allowed for the given value.                                                                 |

#### Return

Boolean

#### Example

```
var isValueLongEnough = defy.minlength('Example', 2)
```



### number(value: Number | String): Boolean

Verify if a given number is an integer or decimal number.

#### Params

| Name                     | Type                     | Description                                                                                                     |
|--------------------------|--------------------------|-----------------------------------------------------------------------------------------------------------------|
| value                    | Number | String          | The value to be validated.                                                                                      |

#### Return

Boolean

#### Example

```
var isValueNumeric = defy.number(123.45)
```



### password(value: String): Boolean

Verify if a given value is a valid password.
A password is considered valid if it is at least 6 characters long and only includes characters from the ASCII printable characters table excluding spaces.

#### Params

| Name                     | Type                     | Description                                                                                                     |
|--------------------------|--------------------------|-----------------------------------------------------------------------------------------------------------------|
| value                    | String                   | The value to be validated.                                                                                      |

#### Return

Boolean

#### Example

```
var isValueValid = defy.password('WowqGY0Gi?oaPnT,MwiJ[kBQ4ID29Hb!')
```



### pattern(value: Number | String, pattern: String): Boolean

Verify if a given value matches a given regular expression pattern.

#### Params

| Name                     | Type                     | Description                                                                                                     |
|--------------------------|--------------------------|-----------------------------------------------------------------------------------------------------------------|
| value                    | Number | String          | The value to be validated.                                                                                      |
| pattern                  | String                   | The regular expression pattern in a format accepted by HTML5 validation rules.                                  |

#### Return

Boolean

#### Example

```
var isValueValid = defy.pattern('GBR', '[A-Za-z]{3}')
```



### required(value: Array | Number | String): Boolean

Verify if a given value is not empty.

#### Params

| Name                     | Type                     | Description                                                                                                     |
|--------------------------|--------------------------|-----------------------------------------------------------------------------------------------------------------|
| value                    | Array | Number | String  | The value to be validated.                                                                                      |

#### Return

Boolean

#### Example

```
var isValueProvided = defy.required('Example')
```



### tel(value: String): Boolean

Verify if a given value is a valid phone number with or without a country code. The digits can be separated by spaces and area codes can be wrapped in parenthesis.

#### Params

| Name                     | Type                     | Description                                                                                                     |
|--------------------------|--------------------------|-----------------------------------------------------------------------------------------------------------------|
| value                    | String                   | The value to be validated. Must be a string to avoid losing digits if a number starts with 0.                   |

#### Return

Boolean

#### Example

```
var isPhoneNumberValid = defy.tel('+44 (0)20 7684 8444');
```



### time(value: String): Boolean

Verify if a given value is a valid time in the format of HH:MM:SS.mm (RFC3339).

#### Params

| Name                     | Type                     | Description                                                                                                     |
|--------------------------|--------------------------|-----------------------------------------------------------------------------------------------------------------|
| value                    | String                   | The value to be validated.                                                                                      |

#### Return

Boolean

#### Example

```
var isTimeValid = defy.time('12:00')
```



### timezone(value: String): Boolean

Verify if a given value is a value timezone offset.

#### Params

| Name                     | Type                     | Description                                                                                                     |
|--------------------------|--------------------------|-----------------------------------------------------------------------------------------------------------------|
| value                    | String                   | The timezone offset value.                                                                                      |

#### Return

Boolean

#### Example

```
var isTimezoneValid = defy.timezone('Z')
```



### url(value: String): Boolean

Verify if a given value is a valid web URL address.

#### Params

| Name                     | Type                     | Description                                                                                                     |
|--------------------------|--------------------------|-----------------------------------------------------------------------------------------------------------------|

CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).


## [v1.1.2] - 2018-01-16
### Fixed

- Fixed the included package files.



## [v1.1.1] - 2018-01-16
### Fixed

- Fixed the path to the main script of the module.



## [v1.1.0] - 2018-01-16
### Added

- It was a bit naive to remove the UMD module distribution. UMD module added back.



## [v1.0.0] - 2018-01-16
### Added

- Added password validation through `defy.password()`

### Fixed

- Fixed an issue in `defy.max()`, `defy.min()` and `defy.required()` that would cause issues when calling them from a function that had its `this` keyword set.

### Breaking changes

#### Removed

- Removed UMD module distribution. Defy now needs to be loaded as an ES6 module.



## [v0.3.0] - 2017-06-22
### Added

- Added age validation based on date of birth through `defy.age()`



## [v0.2.3] - 2017-06-08
### Fixed

- The pattern validation now accepts regular expressions with beggining and end of string tokens.



## [v0.2.2] - 2017-05-31
### Fixed

- The ES module now is transpiled to ES5 for better compatibility and avoiding issues with UglifyJs.


## [v0.2.1] - 2017-05-31
### Fixed

- The documentation now shows the correct function name for `defy.dateExists()`.
- The change log has missing release tag links.


## [v0.2.0] - 2017-05-31
### Added

- Added date value validation through `defy.date()`
- Added date existence validation through `defy.dateExists()`


## [v0.1.3] - 2017-04-30

### Changed

- Changed how the unit test coverage report is sent to Coveralls

### Fixed

- The main Node module now exports the correct object
- Fixed build errors related to the build directory not existing


## [v0.1.2] - 2017-04-29

### Fixed

- Fixed redundant run of tests when publishing the package
- Fixed a package installation error


## [v0.1.1] - 2017-04-29

### Fixed

- Fixed an error when trying to upload the test coverage results to Coveralls which was causing the build to not pass


## [v0.1.0] - 2017-04-29

### Added

- Added a utility function to clean up values by trimming white space through `defy.cleanValue()`
- Added alpha-numeric value validation through `defy.alphanumeric()`
- Added hexadecimal color value validation through `defy.color()`
- Added credit card number value validation through `defy.creditcard()`
- Added email address value validation through `defy.email()`
- Added validation of equality of two values through `defy.match()`
- Added maximum value validation through `defy.max()`
- Added maximum value length validation through `defy.maxlength()`
- Added minimum value validation through `defy.min()`
- Added minimum value length validation through `defy.minlength()`
- Added number value validation through `defy.number()`
- Added regular expression pattern match validation through `defy.pattern()`
- Added phone number value validation through `defy.tel()`
- Added time value validation through `defy.time()`
- Added time zone value validation through `defy.timezone()`
- Added value requirement validation through `defy.required()`
- Added url value validation through `defy.url()`


[Latest]: https://github.com/MrSlide/defy/tree/master
[Unreleased]: https://github.com/MrSlide/defy/tree/develop
[v1.1.2]: https://github.com/MrSlide/defy/tree/v1.1.2
[v1.1.1]: https://github.com/MrSlide/defy/tree/v1.1.1
[v1.1.0]: https://github.com/MrSlide/defy/tree/v1.1.0
[v1.0.0]: https://github.com/MrSlide/defy/tree/v1.0.0
[v0.3.0]: https://github.com/MrSlide/defy/tree/v0.3.0
[v0.2.3]: https://github.com/MrSlide/defy/tree/v0.2.3
[v0.2.2]: https://github.com/MrSlide/defy/tree/v0.2.2
[v0.2.1]: https://github.com/MrSlide/defy/tree/v0.2.1
[v0.2.0]: https://github.com/MrSlide/defy/tree/v0.2.0
[v0.1.3]: https://github.com/MrSlide/defy/tree/v0.1.3
[v0.1.2]: https://github.com/MrSlide/defy/tree/v0.1.2
[v0.1.1]: https://github.com/MrSlide/defy/tree/v0.1.1
[v0.1.0]: https://github.com/MrSlide/defy/tree/v0.1.0


index.js
'use strict'

/**
 * @name Defy.js
 * @desc Small JavaScript library for data validation.
 * @author Luis Rodrigues (http://www.luisrodriguesweb.com)
 * @version 1.1.2
 * @see https://github.com/MrSlide/defy
 * @license MIT
 */

const alphaNumRegEx = /^[0-9a-z]+$/i
const colorRegEx = /^#([0-9a-f]{3}|[0-9a-f]{6})$/i
const emailRegEx = /^[a-zA-Z0-9.+_-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/ // Adapted from https://html.spec.whatwg.org/multipage/forms.html#valid-e-mail-address
const telRegEx = /^(\+[0-9]{1,5})?( ?\([0-9]{1,5}\) ?)?[0-9 -]+$/
const timeRegEx = /^([01]\d|2[0-3]):([0-5]\d):([0-5]\d)(\.[0-9]{2})?$/
const timezoneRegEx = /^(?:Z|[+-](?:1[0-4]|[0][0-9]):(00|30|45))$/
const urlRegEx = /^(?:(?:(?:https?|ftp):)?\/\/)(?:\S+(?::\S*)?@)?(?:(?!(?:10|127)(?:\.\d{1,3}){3})(?!(?:169\.254|192\.168)(?:\.\d{1,3}){2})(?!172\.(?:1[6-9]|2\d|3[0-1])(?:\.\d{1,3}){2})(?:[1-9]\d?|1\d\d|2[01]\d|22[0-3])(?:\.(?:1?\d{1,2}|2[0-4]\d|25[0-5])){2}(?:\.(?:[1-9]\d?|1\d\d|2[0-4]\d|25[0-4]))|(?:(?:[a-z\u00a1-\uffff0-9]-*)*[a-z\u00a1-\uffff0-9]+)(?:\.(?:[a-z\u00a1-\uffff0-9]-*)*[a-z\u00a1-\uffff0-9]+)*(?:\.(?:[a-z\u00a1-\uffff]{2,})).?)(?::\d{2,5})?(?:[/?#]\S*)?$/i
const cardDigitsRegEx = /^[0-9]+([ -]{1}[0-9]+)*$/
const dateRegEx = /^[0-9]{4}-(0[1-9]{1}|1[012]{1})-(0[1-9]{1}|[12]{1}[0-9]{1}|3[01]{1})$/
const passwordRegEx = /^[\x21-\x7E]{6,}$/

/**
 * Parse a valid RFC3339 date string into a `Date` object.
 *
 * @private
 * @since 0.3.0
 * @param {String} value - A valid RFC3339 date string.
 * @return {Date}
 */
function parseDate (value) {
  return new Date(Date.parse(value))
}

/**
 * Get a normalized representation of a value.
 *
 * @private
 * @since 0.1.0
 * @param {*} [value] - Value to be normalized
 * @return {*} - A string representation of the value, or the value itself if a normalized value cannot be achieved
 */
function normalizeValue (value) {
  try {
    return value.toString()
  } catch (e) {
    return value
  }
}

/**
* Clean up a given value to remove extra spaces that may cause false validation results.
*
* @public
* @since 0.1.0
* @param {!Mixed} [value=''] - The value to be validated.
* @return {Boolean}
*
* @example
* var cleanedValue = defy.cleanValue('  Example   ')
*/
export function cleanValue (value = '') {
  return normalizeValue(value).trim().replace(/\s{2,}/g, ' ')
}

/**
 * Verify if someone's age falls within a minimum and maximum age values.
 *
 * @public
 * @since 0.3.0
 * @param {String} dob - A valid RFC3339 date string.
 * @param {Number} [minAge=0] - The minimum accepted age value.
 * @param {Number} [maxAge] - The maximum accepted age value.
 * @return {Boolean|Number} - Returns `false` if the minimum or maximum age are not valid, or the calculated age if valid.
 */
export function age (dob, minAge = 0, maxAge) {
  const now = new Date()
  const nowYear = now.getUTCFullYear()
  const nowMonth = now.getUTCMonth()
  const nowDay = now.getUTCDate()
  const dobDate = parseDate(dob)
  const birthYear = dobDate.getFullYear()
  const birthMonth = dobDate.getMonth()
  const birthDay = dobDate.getDate()
  let age = nowYear - birthYear

  if (nowMonth < birthMonth) {
    age -= 1
  } else if (nowMonth === birthMonth && nowDay < birthDay) {
    age -= 1
  }

  if (minAge && age < minAge) {
    return false
  } else if (maxAge && age > maxAge) {
    return false
  }

  return age
}

/**
* Verify if a given value only contains alpha-numeric characters.
*
* @public
* @since 0.1.0
* @param {!String} value - The value to be validated.
* @return {Boolean}
*
* @example
* var isValueAlphaNumeric = defy.alphaNumeric('Abc123')
*/
export function alphaNumeric (value) {
  return alphaNumRegEx.test(value)
}

/**
* Verify if a given value is a valid hexadecimal format colour with 3 or 6 characters and preceded by the hash sign.
*
* @public
* @since 0.1.0
* @param {!String} value - The value to be validated.
* @return {Boolean}
*
* @example
* var isColourValid = defy.color('#FFFFFF')
*/
export function color (value) {
  return colorRegEx.test(value)
}

/**
* Verify if a given value is a valid credit card number.
*
* @public
* @since 0.1.0
* @param {!String} value - The value to be validated.
* @return {Boolean}
*
* @example
* var isCCValid = defy.creditcard('378282246310005')
*/
export function creditcard (value) {
  // Accept only digits, optionally in groups separated by a single dash or space
  if (!cardDigitsRegEx.test(value)) {
    return false
  }

  value = value.replace(/\D/g, '')

  if (value.length < 12 || value.length > 19) {
    return false
  }

  // Luhn Algorithm
  let nCheck = 0
  let nDigit = 0
  let bEven = false
  let cDigit

  for (let n = value.length - 1; n >= 0; n--) {
    cDigit = value.charAt(n)
    nDigit = parseInt(cDigit, 10)

    if (bEven) {
      if ((nDigit *= 2) > 9) nDigit -= 9
    }

    nCheck += nDigit
    bEven = !bEven
  }

  return (nCheck % 10) === 0
}
/**
 * Verify if a given value is a valid date in the format YYYY-MM-DD.
 *
 * @public
 * @since 0.2.0
 * @param {!String} value - The value to be validated.
 * @return {Boolean}
 *
 * @example
 * var isDateValid = defy.date('2016-01-31')
 */
export function date (value) {
  return dateRegEx.test(value)
}

/**
 * Verify if a given value is an existing date.
 *
 * @public
 * @since 0.2.0
 * @param {String} value - A valid RFC3339 date string.
 * @return {Boolean}
 *
 * @example
 * var dateExists = defy.dateExists('2017-05-31')
 */
export function dateExists (value) {
  const dateObj = parseDate(value)

  return !isNaN(dateObj.getTime())
}

/**
* Verify if a given value is a valid email address.
*
* @public
* @since 0.1.0
* @param {!String} value - The value to be validated.
* @return {Boolean}
*
* @example
* var isEmailValid = defy.email('user@world.net');
*/
export function email (value) {
  const [localPart] = value.split('@')

  // Validate value and part lengths
  if (localPart.length > 64) {
    return false
  }

  return emailRegEx.test(value)
}

/**
* Verify if two given values match.
* Both values will be treated as strings.
*
* @public
* @since 0.1.0
* @param {Mixed} valueA - The first value to match.
* @param {Mixed} valueB - The second value, which will be compared with the first.
* @return {Boolean}
*
* @example
* var doValuesMatch = defy.match(1, '1')
*/
export function match (valueA, valueB) {
  // Normalize values into strings
  const valA = normalizeValue(valueA)
  const valB = normalizeValue(valueB)

  if (valA !== valB) {
    return false
  }

  return true
}

/**
* Verify if a given value does not exceed another given value.
* Both values will be treated as strings.
*
* @public
* @since 0.1.0
* @param {!Number|String} value - The value to be validated.
* @param {!Number|String} maxValue - The maximum allowed for the given value.
* @return {Boolean}
*
* @example
* var isValueLessThanEleven = defy.max(5, 10)
*/
export function max (value, maxValue) {
  let val = value
  let maxVal = maxValue

  if (normalizeValue(val).length !== normalizeValue(maxVal).length) {
    if (number(val)) {
      val = parseFloat(val)
    }

    if (number(maxVal)) {
      maxVal = parseFloat(maxVal)
    }
  }

  if (val > maxVal) {
    return false
  }

  return true
}

/**
* Verify if a given value's length does not exceed the given maximum length.
* For strings this will validate the number of characters.
* For numbers, this will validate the number of digits, including the decimal places, and the decimal mark.
*
* @public
* @since 0.1.0
* @param {!Number|String} value - The value to be validated.
* @param {!Number} maxLen - The maximum length allowed for the given value.
* @return {Boolean}
*
* @example
* var isValueNotTooLong = defy.maxlength('Example', 10)
*/
export function maxlength (value, maxLen) {
  let val = value

  if (typeof value === 'number') {
    val = normalizeValue(value)
  }

  const mxLength = parseInt(maxLen, 10)

  if (val.length > mxLength) {
    return false
  }

  return true
}

/**
* Verify if a given value is not inferior to another given value.
* Both values will be treated as strings.
*
* @public
* @since 0.1.0
* @param {!Number|String} value - The value to be validated.
* @param {!Number|String} minValue - The minimum allowed for the given value.
* @return {Boolean}
*
* @example
* var isValueGreaterThanZero = defy.min(5, 0)
*/
export function min (value, minValue) {
  let val = value
  let minVal = minValue

  if (normalizeValue(val).length !== normalizeValue(minVal).length) {
    if (number(val)) {
      val = parseFloat(val)
    }

    if (number(minVal)) {
      minVal = parseFloat(minVal)
    }
  }

  if (val < minVal) {
    return false
  }

  return true
}

/**
* Verify if a given value's length is at least the given minimum length.
* For strings this will validate the number of characters.
* For numbers, this will validate the number of digits, including the decimal places, and the decimal mark.
*
* @public
* @since 0.1.0
* @param {!Number|String} value - The value to be validated.
* @param {!Number} minLen - The minimum length allowed for the given value.
* @return {Boolean}
*
* @example
* var isValueLongEnough = defy.minlength('Example', 2)
*/
export function minlength (value, minLen) {
  let val = value

  if (typeof value === 'number') {
    val = normalizeValue(value)
  }

  const mnLength = parseInt(minLen, 10)

  if (val.length < mnLength) {
    return false
  }

  return true
}

/**
* Verify if a given number is an integer or decimal number.
*
* @public
* @since 0.1.0
* @param {!Number|String} value - The value to be validated.
* @return {Boolean}
*
* @example
* var isValueNumeric = defy.number(123.45)
*/
export function number (value) {
  return !isNaN(parseFloat(normalizeValue(value))) && isFinite(value)
}

/**
 * Verify if a password has only valid characters and has a minimum length of 6 characters.
 * The valid characters are the ASCII printable characters, excluding the space character.
 *
 * @public
 * @since 1.0.0
 * @param {!String} value - The value to be validated.
 * @return {Boolean}
 *
 * @example
 * var isPasswordValid = defy.password('WowqGY0Gi?oaPnT,MwiJ[kBQ4ID29Hb!')
 */
export function password (value) {
  return passwordRegEx.test(value)
}

/**
* Verify if a given value matches a given regular expression pattern.
*
* @public
* @since 0.1.0
* @param {!Number|String} value - The value to be validated.
* @param {!String} pattern - The regular expression pattern in a format accepted by HTML5 validation rules.
* @return {Boolean}
*
* @example
* var isValueValid = defy.pattern('GBR', '[A-Za-z]{3}')
*/
export function pattern (value, pattern) {
  const regEx = new RegExp(pattern)

  return regEx.test(value)
}

/**
* Verify if a given value is a valid phone number with or without a country code.
* The digits can be separated by spaces and area codes can be wrapped in parenthesis.
*
* @public
* @since 0.1.0
* @param {!String} value - The value to be validated. Must be a string to avoid losing digits if a number starts with 0.
* @return {Boolean}
*
* @example
* var isPhoneNumberValid = defy.tel('+44 (0)20 7684 8444');
*/
export function tel (value) {
  if (value.length > 20) {
    return false
  }

  return telRegEx.test(value)
}

/**
* Verify if a given value is a valid time in the format of HH:MM:SS.mm ({@link RFC3339}).
*
* @public
* @since 0.1.0
* @param {!String} value - The value to be validated.
* @return {Boolean}
*
* @example
* var isTimeValid = defy.time('12:00')
*/
export function time (value) {
  return timeRegEx.test(value)
}

/**
* Verify if a given value is a value timezone offset.
*
* @public
* @since 0.1.0
* @param {!String} value - The timezone offset value
* @return {Boolean}
*
* @example
* var isTimezoneValid = defy.timezone('Z')
*/
export function timezone (value) {
  return timezoneRegEx.test(value)
}

/**
* Verify if a given value is not empty.
*
* @public
* @since 0.1.0
* @param {!Array|Number|String} value - The value to be validated.
* @return {Boolean}
*
* @example
* var isValueProvided = defy.required('Example')
*/
export function required (value) {
  let inputValue = value

  // Clean up a string before validating it
  if (typeof value === 'string') {
    inputValue = cleanValue(inputValue)
  }

  if (inputValue === undefined || inputValue === null) {
    return false
  } else if (typeof inputValue !== 'number' && inputValue.length === 0) {
    return false
  }

  return true
}

/**
* Verify if a given value is a valid web URL address.
*
* @public
* @since 0.1.0
* @param {!String} value - The value to be validated.
* @return {Boolean}
*
* @example
* var isUrlValid = defy.url('https://www.google.com')
*/
export function url (value) {
  return urlRegEx.test(value)
}
